### PR TITLE
Added /slimefun drop command

### PR DESCRIPTION
**Pull request by NCBPFluffyBear**
## Description
Adds `/sf drop`, a command which drops a Slimefun item with a selectable quantity at the specified coordinates. This is helpful for scripted actions, such as mob drops.

## Changes
Added the `DropCommand` subcommand and registered under subcommands
Added `drop-item` to `messages_en.yml`
Added `DropCommandTest` unit test

## Related Issues
This was suggested on Discord so that MythicMobs could drop Slimefun items

## Checklist

- [x] I have fully tested the proposed changes and promise that they will not break everything into chaos.
- [x] I have also tested the proposed changes in combination with various popular addons and can confirm my changes do not break them.
- [x] I followed the existing code standards and didn't mess up the formatting.
- [x] I did my best to add documentation to any public classes or methods I added.
- [ ] I have added `Nonnull` and `Nullable` annotations to my methods to indicate their behaviour for null values
- [x] I added sufficient Unit Tests to cover my code.

### CODE REVIEW

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/DropCommand.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    }
+
+    @Override
+    public void onExecute(CommandSender sender, String[] args) {
+        if (!(sender instanceof Player) || sender.hasPermission("slimefun.cheat.items")) {
+            if (args.length > 2) {
+                if (args.length > 5) {
+                    SlimefunItem sfItem = SlimefunItem.getByID(args[5].toUpperCase(Locale.ROOT));
+
+                    if (sfItem != null) {
```
Comment (svr333) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/DropCommand.java:45]: I myself am not a big fan of this many nested ifs; I usually invert them and then return if the condition is not met.
Dont know how it is handled around here though.

**Review by svr333 — COMMENTED** [review]
Two really minor remarks.
This doesnt have too many usecases but it can be really damn useful in certain areas.

File: src/test/java/io/github/thebusybiscuit/slimefun4/testing/tests/commands/DropCommandTest.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+import io.github.thebusybiscuit.slimefun4.implementation.SlimefunPlugin;
+import io.github.thebusybiscuit.slimefun4.testing.TestUtilities;
+import me.mrCookieSlime.Slimefun.Lists.RecipeType;
+import me.mrCookieSlime.Slimefun.Objects.Category;
+import me.mrCookieSlime.Slimefun.Objects.SlimefunItem.SlimefunItem;
+import me.mrCookieSlime.Slimefun.api.SlimefunItemStack;
+
+import be.seeseemelk.mockbukkit.MockBukkit;
+import be.seeseemelk.mockbukkit.ServerMock;
+import org.junit.jupiter.api.*;
```
Comment (poma123) [inline src/test/java/io/github/thebusybiscuit/slimefun4/testing/tests/commands/DropCommandTest.java:17]: Please don't use star imports

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/DropCommand.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+    DropCommand(SlimefunPlugin plugin, SlimefunCommand cmd) {
+        super(plugin, cmd, "drop", false);
+    }
+
+    @Override
+    public void onExecute(CommandSender sender, String[] args) {
+        if (!(sender instanceof Player) || sender.hasPermission("slimefun.cheat.items")) {
+            if (args.length > 2) {
+                if (args.length > 5) {
```
Comment (poma123) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/DropCommand.java:42]: I'd rather check whether `args.length` is greater than 5 here, no need to check it twice

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/DropCommand.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+import io.github.thebusybiscuit.cscorelib2.item.CustomItem;
+import io.github.thebusybiscuit.slimefun4.core.commands.SlimefunCommand;
+import io.github.thebusybiscuit.slimefun4.core.commands.SubCommand;
+import io.github.thebusybiscuit.slimefun4.core.multiblocks.MultiBlockMachine;
+import io.github.thebusybiscuit.slimefun4.implementation.SlimefunPlugin;
+import io.github.thebusybiscuit.slimefun4.utils.PatternUtils;
+import me.mrCookieSlime.Slimefun.Objects.SlimefunItem.SlimefunItem;
+
+/**
+ * The drop command drops a {@link SlimefunItem} of
```
Comment (poma123) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/DropCommand.java:19]:
```suggestion
 * {@link DropCommand} drops a {@link SlimefunItem} of
```

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/DropCommand.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+        if (sfItem instanceof MultiBlockMachine) {
+            SlimefunPlugin.getLocalization().sendMessage(sender, "guide.cheat.no-multiblocks");
+        } else {
+            int amount = parseAmount(args);
+
+            if (amount > 0) {
+                Bukkit.getWorld(world).dropItem(new Location(Bukkit.getWorld(world), Integer.parseInt(x), Integer.parseInt(y), Integer.parseInt(z)), new CustomItem(sfItem.getItem(), amount));
+                SlimefunPlugin.getLocalization().sendMessage(sender, "messages.drop-item", true, msg -> msg.replace(PLACEHOLDER_WORLD, args[1]).replace(PLACEHOLDER_X, args[2]).replace(PLACEHOLDER_Y, args[3]).replace(PLACEHOLDER_Z, args[4]).replace(PLACEHOLDER_ITEM, sfItem.getItemName()).replace(PLACEHOLDER_AMOUNT, String.valueOf(amount)));
+            } else {
+                SlimefunPlugin.getLocalization().sendMessage(sender, "messages.not-valid-amount", true, msg -> msg.replace(PLACEHOLDER_AMOUNT, args[3]));
```
Comment (poma123) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/DropCommand.java:69]:
```suggestion
                SlimefunPlugin.getLocalization().sendMessage(sender, "messages.invalid-amount", true, msg -> msg.replace(PLACEHOLDER_AMOUNT, args[3]));
```

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/DropCommand.java [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    public void onExecute(CommandSender sender, String[] args) {
+        if (!(sender instanceof Player) || sender.hasPermission("slimefun.cheat.items")) {
+            if (args.length > 2) {
+                if (args.length > 5) {
+                    SlimefunItem sfItem = SlimefunItem.getByID(args[5].toUpperCase(Locale.ROOT));
+
+                    if (sfItem != null) {
+                        dropItem(sender, args[1], args[2], args[3], args[4], sfItem, args);
+                    } else {
+                        SlimefunPlugin.getLocalization().sendMessage(sender, "messages.not-valid-item", true, msg -> msg.replace(PLACEHOLDER_ITEM, args[2]));
```
Comment (poma123) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/DropCommand.java:48]:
```suggestion
                        SlimefunPlugin.getLocalization().sendMessage(sender, "messages.invalid-item", true, msg -> msg.replace(PLACEHOLDER_ITEM, args[2]));
```

**Review by poma123 — CHANGES_REQUESTED** [review]

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/DropCommand.java [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+            }
+        }
+    }
+
+    private int parseAmount(String[] args) {
+        int amount = 1;
+
+        if (args.length == 7) {
+            if (PatternUtils.NUMERIC.matcher(args[6]).matches()) {
+                amount = Integer.parseInt(args[6]);
```
Comment (J3fftw1) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/DropCommand.java:79]: These 2 could be combined with &&

Comment (J3fftw1) [reply src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/DropCommand.java:79]: The ifs i mean*

Comment (NCBPFluffyBear) [reply src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/DropCommand.java:79]: cant, one returns 0 the other returns 1

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/SlimefunTabCompleter.java [hunk 8]
```diff
@@ -18,6 +21,9 @@
 import io.github.thebusybiscuit.slimefun4.implementation.SlimefunPlugin;
 import me.mrCookieSlime.Slimefun.Objects.SlimefunItem.SlimefunItem;
 
+import com.google.common.base.Functions;
+import com.google.common.collect.Lists;
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/SlimefunTabCompleter.java:25]: These are not used, remove

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/DropCommand.java [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+ * @author TheBusyBiscuit (Based off of GiveCommand)
+ */
+class DropCommand extends SubCommand {
+
+    private static final String PLACEHOLDER_WORLD = "%world%";
+    private static final String PLACEHOLDER_X = "%x%";
+    private static final String PLACEHOLDER_Y = "%y%";
+    private static final String PLACEHOLDER_Z = "%z%";
+    private static final String PLACEHOLDER_ITEM = "%item%";
+    private static final String PLACEHOLDER_AMOUNT = "%amount%";
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/DropCommand.java:33]: Always hated %% as a placeholder prefix/suffix, I prefer {} but that's my preference. {world} just looks better and for me is more readable than %world%.

Others opinions?

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/DropCommand.java [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+
+    private void dropItem(CommandSender sender, String world, String x, String y, String z, SlimefunItem sfItem, String[] args) {
+        if (sfItem instanceof MultiBlockMachine) {
+            SlimefunPlugin.getLocalization().sendMessage(sender, "guide.cheat.no-multiblocks");
+        } else {
+            int amount = parseAmount(args);
+
+            if (amount > 0) {
+                Bukkit.getWorld(world).dropItem(new Location(Bukkit.getWorld(world), Integer.parseInt(x), Integer.parseInt(y), Integer.parseInt(z)), new CustomItem(sfItem.getItem(), amount));
+                SlimefunPlugin.getLocalization().sendMessage(sender, "messages.drop-item", true, msg -> msg.replace(PLACEHOLDER_WORLD, args[1]).replace(PLACEHOLDER_X, args[2]).replace(PLACEHOLDER_Y, args[3]).replace(PLACEHOLDER_Z, args[4]).replace(PLACEHOLDER_ITEM, sfItem.getItemName()).replace(PLACEHOLDER_AMOUNT, String.valueOf(amount)));
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/DropCommand.java:67]: This line is huge, please format nicely

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/DropCommand.java [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+    }
+
+    private void dropItem(CommandSender sender, String world, String x, String y, String z, SlimefunItem sfItem, String[] args) {
+        if (sfItem instanceof MultiBlockMachine) {
+            SlimefunPlugin.getLocalization().sendMessage(sender, "guide.cheat.no-multiblocks");
+        } else {
+            int amount = parseAmount(args);
+
+            if (amount > 0) {
+                Bukkit.getWorld(world).dropItem(new Location(Bukkit.getWorld(world), Integer.parseInt(x), Integer.parseInt(y), Integer.parseInt(z)), new CustomItem(sfItem.getItem(), amount));
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/DropCommand.java:66]: There should be checks here or above that these are actually ints. `PatternUtils.NUMERIC.matcher(s).matches()`

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/DropCommand.java [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+                SlimefunPlugin.getLocalization().sendMessage(sender, "messages.drop-item", true, msg -> msg.replace(PLACEHOLDER_WORLD, args[1]).replace(PLACEHOLDER_X, args[2]).replace(PLACEHOLDER_Y, args[3]).replace(PLACEHOLDER_Z, args[4]).replace(PLACEHOLDER_ITEM, sfItem.getItemName()).replace(PLACEHOLDER_AMOUNT, String.valueOf(amount)));
+            } else {
+                SlimefunPlugin.getLocalization().sendMessage(sender, "messages.invalid-amount", true, msg -> msg.replace(PLACEHOLDER_AMOUNT, args[3]));
+            }
+        }
+    }
+
+    private int parseAmount(String[] args) {
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/DropCommand.java:74]: not sure why you pass an array here. Just pass a single string, can even use this for the x,y,z above.

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/DropCommand.java [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+
+        if (args.length == 7) {
+            if (PatternUtils.NUMERIC.matcher(args[6]).matches()) {
+                amount = Integer.parseInt(args[6]);
+            } else {
+                return 0;
+            }
+        }
+
+        return amount;
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/DropCommand.java:85]:
```suggestion
if (PatternUtils.NUMERIC.matcher(s).matches()) {
    return OptionalInt.of(Integer.parseInt(s));
} else {
    return OptionalInt.empty();
}
```

File: src/main/resources/languages/messages_en.yml [hunk 14]
```diff
@@ -132,7 +134,6 @@ messages:
   no-tome-yourself: '&cYou cannot use the &4Tome of Knowledge &con yourself...'
   multimeter: '&bStored Energy: &3%stored% &b/ &3%capacity%'
   piglin-barter: '&4You cannot barter with piglins using Slimefun items'
-  bee-suit-slow-fall: '&eYour Bee Wings will help you to get back to the ground safe and slow'
```
Comment (WalshyDev) [inline src/main/resources/languages/messages_en.yml:135]: don't remove this

File: src/test/java/io/github/thebusybiscuit/slimefun4/testing/tests/commands/DropCommandTest.java [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+    void testCommand() {
+        Category category = TestUtilities.getCategory(plugin, "testcategory");
+        SlimefunItemStack TEST_ITEM = new SlimefunItemStack("DROP_COMMAND_TEST_ITEM", Material.REDSTONE_BLOCK, "Drop Test Item");
+        new SlimefunItem(category, TEST_ITEM, RecipeType.NULL, new ItemStack[9]).register(plugin);
+        server.addSimpleWorld("testworld");
+
+        server.executeConsole("slimefun", "drop", "testworld", "0", "0", "0", "DROP_COMMAND_TEST_ITEM", "1");
+        for (Entity en : server.getWorld("testworld").getEntities()) {
+            Assertions.assertEquals(SlimefunItem.getByItem(((Item) en).getItemStack()), TEST_ITEM.getItem());
+        }
```
Comment (WalshyDev) [inline src/test/java/io/github/thebusybiscuit/slimefun4/testing/tests/commands/DropCommandTest.java:50]: There also should be tests to make sure validation works, so non-extistent world, invalid coords, invalid item, too high amount/0, etc

**Review by WalshyDev — CHANGES_REQUESTED** [review]

**Comment by TheBusyBiscuit** [comment]
Closing this since there hasn't been a response in over a month now.

**Pull request closed without merging**